### PR TITLE
Add type annotations to the python codebase

### PR DESCRIPTION
**Pull request by youtux**
### 🤔 What's changed?

* Added type annotations to the python codebase.
* Added `py.typed` file marker

There are still quite some errors (actual bug, mypy idiosyncrasies, ...) , but at least all the signatures are annotated.

### ⚡️ What's your motivation? 

Static type checking make things safer and easier to find usages.

### 🏷️ What kind of change is this?

- :bank: Refactoring/debt/DX (improvement to code design, tooling, etc. without changing behaviour)

### 📋 Checklist:

- [x] I agree to respect and uphold the [Cucumber Community Code of Conduct](https://github.com/cucumber/.github/tree/main?tab=coc-ov-file)
- [ ] I've changed the behaviour of the code
  - [ ] I have added/updated tests to cover my changes.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.
- [x] Users should know about my change
  - [x] I have added an entry to the "Unreleased" section of the [**CHANGELOG**](../blob/main/CHANGELOG.md), linking to this pull request.

----

*This text was originally generated from a [template](https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/about-issue-and-pull-request-templates), then edited by hand. [You can modify the template here.](https://github.com/cucumber/.github/edit/main/.github/PULL_REQUEST_TEMPLATE.md)*

### CODE REVIEW

**Comment by jsa34** [comment]
Perhaps instead of using typing.List instead of list for the typings, given I guess that we are so close to EOL for py3.8, should we just go for 3.9+?

**Comment by youtux** [comment]
> Perhaps instead of using typing.List instead of list for the typings, given I guess that we are so close to EOL for py3.8, should we just go for 3.9+?

I fixed those lines, it was no big deal, so let's keep 3.8 until it goes EOL

File: python/test/gherkin_in_markdown_token_matcher_test.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 
 def test_it_matches_FeatureLine():
     tm = GherkinInMarkdownTokenMatcher('en')
-    line = GherkinLine('''## Feature: hello''',location['line'])
```
Comment (jsa34) [inline python/test/gherkin_in_markdown_token_matcher_test.py:11]: Maybe I'm missing something here?  I'm not sure this test has this change?

Comment (youtux) [reply python/test/gherkin_in_markdown_token_matcher_test.py]: nice catch! not sure how this happened... Fixed in abed0db7f255cb95deb92699c4d5590abddafe22

**Review by jsa34 — COMMENTED** [review]
Just my initial spots

File: python/gherkin/parser_types.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+class Feature(TypedDict):
+    location: Location
+    tags: list[Tag]
+    language: str
+    keyword: str
+    name: str
+    description: str
+    children: list[Container]
+
+class GherkinDocument(TypedDict):
```
[the remark this answers is not in the document]
Comment (jenisys) [reply python/gherkin/parser_types.py:106]: Mmh, I am just curious of the naming scheme related to `Container` and its variants:

* Normally, a `Container` contains many items, but here it contains just one item (which surprises me)
* Therefore, `Container` should be better named `Item`, `ModelItem`, `Envelope` or …

[the remark this answers is not in the document]
Comment (youtux) [reply python/gherkin/parser_types.py:106]: > Shouldn't this have an optional field of "uri", from looking at the tests?

well, the result of `Parser.parse()` does not include a `uri`, and places that require it use the subclass `GherkinDocumentWithURI`.

[the remark this answers is not in the document]
Comment (youtux) [reply python/gherkin/parser_types.py:106]: > Mmh, I am just curious of the naming scheme related to `Container` and its variants:
> 
> * Normally, a `Container` contains many items, but here it contains just one item (which surprises me)
> * Therefore, `Container` should be better named `Item`, `ModelItem`, `Envelope` or …

I like the Envelope naming, will use that. Changed in 6e719e34

[the remark this answers is not in the document]
Comment (youtux) [reply python/gherkin/parser_types.py:106]: sorry, wrong branch. Updated actually in f95ab9a1

File: python/gherkin/ast_builder.py [hunk 3]
```diff
@@ -71,20 +94,42 @@ def ensure_cell_count(self, rows):
...
+        | list[TableRow]
+        | str
+        | None
+        | Rule
+        | Feature
+        | GherkinDocument
+        | AstNode
+    ):
         if node.rule_type == 'Step':
-            step_line = node.get_token('StepLine')
```
[the remark this answers is not in the document]
Comment (youtux) [reply python/gherkin/ast_builder.py:87]: > Shouldn't the cast happen in get_token(), as the type hint and expectation is that a object of type Token will be returned?

good one, `get_token` was already annotated to return a `Token`, so this cast was unnecessary. 
Removed in 3cda2fb4

> Also, the Token type hint doesn't acknowledge the attributes set when a token is matched and created (e.g. matched_keyword)

Yep, that's a problem of the codebase, I didn't want to fix the codebase in this PR, just add type annotations.

In a next PR we can fix all these issues

**Review by jsa34 — COMMENTED** [review]
Looks good to me.

**Pull request merged**